File: Android/AnUitorService/src/com/scurab/android/anuitor/hierarchy/IdsHelper.java
package com.scurab.android.anuitor.hierarchy;

import android.content.res.Resources;
import android.util.Log;
import android.util.SparseArray;
import android.util.TypedValue;

import com.google.gson.Gson;
import com.scurab.android.anuitor.model.Pair;
import com.scurab.android.anuitor.model.Tuple;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * User: jbruchanov
 * Date: 12/05/2014
 * Time: 13:52
 */
public class IdsHelper {

    public enum RefType {
        anim, animator, array, attr,
        bool,
        color,
        dimen, drawable,
        fraction,
        id, integer, interpolator,
        layout,
        menu, mipmap,
        plurals,
        raw,
        string, style, styleable,
        unknown,
        xml;
    }

    private static final HashMap<String, SparseArray<String>> VALUES;

    static {
        VALUES = new HashMap<String, SparseArray<String>>();
    }

    public static void loadValues(Class<?> Rclass) throws NoSuchFieldException, ClassNotFoundException {
        if (!VALUES.isEmpty()) {
            return;
        }
        Class<?>[] clzs = Rclass.getClasses();
        for (Class<?> clz : clzs) {
            fillClass(clz, false);
        }

        clzs = android.R.class.getClasses();
        for (Class<?> clz : clzs) {
            fillClass(clz, true);
        }
    }

    private static void fillClass(Class<?> containerClass, boolean android) {
        SparseArray<String> values = new SparseArray<String>();
        String name = containerClass.getSimpleName();
        String v = containerClass.getCanonicalName();
        if (!android) {
            v = v.replace(containerClass.getPackage().getName(), "").substring(1);
        }
        VALUES.put(v, values);
        fillFields(name, values, containerClass.getFields(), android);
    }

    private static void fillFields(String type, SparseArray<String> container, Field[] fields, boolean android) {
        for (Field field : fields) {
            field.setAccessible(true);
            if (field.getType() == int.class) {
                String name = field.getName();
                try {
                    int value = field.getInt(null);
                    container.put(value, String.format("@%s/%s", android ? "android:" + type : type, name));
                } catch (Exception e) {
                    //this should never happen if we have setAccessible
                    continue;
                }
            }
        }
    }

    private static Class<?> getClass(Class<?> parent, String name) {
        Class<?>[] classes = parent.getClasses();
        for (Class<?> aClass : classes) {
            if (name.equals(aClass.getSimpleName())) {
                return aClass;
            }
        }
        return null;
    }

    public static String getNameForId(int id) {
        if (id == -1) {
            return "undefined";
        }
        for (String type : VALUES.keySet()) {
            SparseArray<String> array = VALUES.get(type);
            String result = array.get(id);
            if (result != null) {
                return result;
            }
        }
        return null;
    }

    public static RefType getType(int id) {
        for (String type : VALUES.keySet()) {
            SparseArray<String> array = VALUES.get(type);
            String result = array.get(id, null);
            if (result != null) {
                String[] values = type.split("\\.");
                String value = values[values.length - 1];
                RefType refType = RefType.valueOf(value);
                return refType;
            }
        }
        return RefType.unknown;
    }

    public static String toJson() {
        return toJson(null);
    }

    public static String toJson(Resources res) {
        TypedValue tv = new TypedValue();
        CharSequence location = null;
        HashMap<String, List<Pair<Object, Object>>> result = new HashMap<String, List<Pair<Object, Object>>>();
        for (String type : VALUES.keySet()) {
            boolean showValue = res != null && ((type.equals("R.drawable") || type.equals("R.layout") || type.equals("R.color")));
            List<Pair<Object, Object>> list = new ArrayList<Pair<Object, Object>>();
            result.put(type, list);
            SparseArray<String> sa = VALUES.get(type);
            for (int i = 0, n = sa.size(); i < n; i++) {
                int key = sa.keyAt(i);
                String value = sa.get(key);
                if (showValue) {
                    try {
                        res.getValue(key, tv, false);
                        location = tv.string;
                    } catch (Exception e) {
                        //just for sure
                        Log.e("IdsHelper", String.format("Name:%s Err:%s", value, e.getMessage()));
                        e.printStackTrace();
                    }
                }
                list.add(new Tuple<Object, Object, Object>(key, value, showValue ? location : null));
                location = null;
            }
        }
        return new Gson().toJson(result);
    }
}

File: uitor-android/uitor-service/src/main/core/com/scurab/android/uitor/hierarchy/IdsHelper.kt
package com.scurab.android.uitor.hierarchy

import android.content.Context
import android.content.res.Resources
import android.util.Log
import android.util.TypedValue
import java.lang.reflect.Field

enum class RefType {
    anim, animator, array, attr,
    bool,
    color,
    dimen, drawable,
    fraction, font,
    id, integer, interpolator,
    layout,
    menu, mipmap,
    navigation,
    plurals,
    raw,
    string, style, styleable,
    transition,
    unknown,
    xml
}

/**
 * Class to help you with android IDs
 */
object IdsHelper {

    internal var appResources: Resources? = null

    @JvmStatic
    internal var data = mutableMapOf<String, Map<Int, String>>()

    @JvmStatic
    var RClass: Class<*>? = null

    /**
     * Preload names of IDs defined in the app
     */
    @JvmStatic
    fun loadValues(context: Context, Rclass: Class<*>) {
        if (data.isNotEmpty()) {
            data.clear()
        }
        appResources = context.applicationContext.resources
        Rclass.classes.forEach { fillClass(it, false) }
        android.R::class.java.classes.forEach { fillClass(it, true) }
        RClass = Rclass
    }

    /**
     * Returns name for particular id<br/>
     * If id is not defined {@link android.view.View#getId()} returns -1, this method returns "undefined".
     * If the id is not found (maybe android.internal stuff) null is returned.
     * @param id
     * @return
     */
    @JvmStatic
    fun getNameForId(id: Int): String {
        if (id == -1) {
            return "View.NO_ID"
        }
        return data.values
            .firstOrNull { it.contains(id) }
            ?.let { it[id] }
            ?: kotlin.runCatching { appResources?.getResourceName(id) }.getOrNull()
            ?: kotlin.runCatching { Resources.getSystem().getResourceName(id) }.getOrNull()
            ?: id.toString()
    }

    /**
     * Get type of particular id. [com.scurab.android.uitor.hierarchy.IdsHelper.RefType.unknown] is returned if not found.
     * @param id
     * @return
     */
    @JvmStatic
    fun getType(id: Int): RefType {
        return data
            .filter { it.value.containsKey(id) }
            .entries.firstOrNull()
            ?.let { (type, _) -> type.split(".").lastOrNull() }
            ?.let { RefType.valueOf(it) }
            ?: RefType.unknown
    }

    /**
     * Get all resources for serialization
     */
    @JvmStatic
    fun getAllResources(res: Resources): Map<String, List<ResourceDTO>> {
        val tv = TypedValue()
        return data.mapValues { (group, values) ->
            val showValue = group == "R.drawable" || group == "R.layout" || group == "R.color"
            values.map { (resId, name) ->
                val location = if (showValue) {
                    res.getValue(resId, tv)
                } else null
                ResourceDTO(resId, name, location)
            }
        }
    }

    private fun fillClass(containerClass: Class<*>, vanillaAndroid: Boolean) {
        val values = mutableMapOf<Int, String>()
        val name = containerClass.simpleName
        var v = containerClass.canonicalName
            ?: throw NullPointerException("containerClass.canonicalName is null?!")
        val pckg = containerClass.`package`?.name
        if (!vanillaAndroid) {
            require(pckg != null) { "Containerclass.package:$containerClass has null name ?!" }
            v = v.replace(pckg, "").substring(1)
        }
        data[v] = values
        fillFields(name, values, containerClass.fields, vanillaAndroid)
    }

    private fun fillFields(
        type: String,
        container: MutableMap<Int, String>,
        fields: Array<Field>,
        vanillaAndroid: Boolean
    ) {
        for (field in fields) {
            field.isAccessible = true
            if (field.type == Int::class.javaPrimitiveType) {
                val name = field.name
                try {
                    val key = field.getInt(null)
                    val tag = if ("attr" == type) "?" else "@"
                    val prefix = if (vanillaAndroid) "android:$type" else type
                    container[key] = "$tag$prefix/$name"
                } catch (e: Throwable) { // this should never happen if we have setAccessible
                    continue
                }
            }
        }
    }

    private fun Resources.getValue(resId: Int, typedValue: TypedValue): CharSequence? {
        return try {
            getValue(resId, typedValue, false)
            typedValue.string
        } catch (e: Throwable) {
            // just for sure
            Log.e("IdsHelper", "Name:${getResourceName(resId)} Err:${e.message ?: "Null"}")
            e.printStackTrace()
            null
        }
    }
}

class ResourceDTO(
    val key: Int,
    val value: String,
    val contextValue: CharSequence?
)
